experiments/process_sandbox/filetree.h

```diff
@@ -0,0 +1,319 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+/**
+ * This file contains the abstractions required for a lightweight virtual
+ * filesystem exported to a sandbox.  The code outside the sandbox is
+ * responsible for defining the structure of a file tree and providing either
+ * handles to directories in a real filesystem or to individual file objects.
+ *
+ * In a Capsicum world, this can mostly live entirely inside the sandbox
+ * because the `*at()` calls are safe.  If sandboxes are implemented with
+ * system-call filtering, this must live outside the sandbox and provide
+ * handles inside.  In theory, sandboxes could be implemented using a jail-like
+ * mechanism and in the VFS would be replaced with a real directory tree
+ * containing nullfs mounts and hard links.
+ *
+ * TODO: Eventually, we should add permissions masks on entities in this
+ * filesystem.
+ */
+
+#pragma once
+#include "platform/platform.h"
+
+#include <optional>
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <variant>
+
+namespace sandbox
+{
+  /**
+   * Class encapsulating a filesystem exported to a sandbox.  This is a very
+   * lightweight simulation of a real filesystem.  This provides a directory
+   * structure where each leaf node is either a directory handle to a real
+   * directory or a file handle to a file (or something equivalent, such as
+   * an anonymous shared memory object).
+   *
+   * Note: The directories in this structure that are not represented by
+   * handles cannot be opened or otherwise inspected in the sandbox.  If the
+   * tree stores `/tmp/example/somefile`, where only `somefile` is
+   * represented by a real file descriptor, the sandbox will not see `/tmp`
+   * or `/tmp/example`.  This restriction may be lifted in the future if it
+   * causes significant problems.
```

Comment:
Is there a reason for this, other than implementation complexity? Or is this restricted to avoid the sandbox from iterating the file system like `cat ../../var/log/security` without a proper permission management?